apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/TimedBufferingConsumer.cs

```diff
@@ -61,37 +63,88 @@ private TimedBufferingConsumer(IConsumer<T> consumer, TimeSpan waitTime, IClock
         /// <param name="waitTime">The minimum amount of time between automatic flushes.</param>
         /// <param name="clock">A clock for getting the current timestamp.</param>
         public static TimedBufferingConsumer<T> Create(IConsumer<T> consumer, TimeSpan waitTime, IClock clock = null)
+            => new TimedBufferingConsumer<T>(consumer, waitTime, clock ?? SystemClock.Instance);
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
         {
-            return new TimedBufferingConsumer<T>(consumer, waitTime, clock ?? SystemClock.Instance);
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            _semaphore.Wait();
+            _items.AddRange(items);
+            if (_clock.GetCurrentDateTimeUtc() >= _nextFlush)
+            { 
+                Flush(waitForSemaphore: false);
+            }
+            _semaphore.Release();
         }
 
         /// <inheritdoc />
-        public void Receive(IEnumerable<T> items)
+        public async Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            lock (_mutex)
+            try
             {
+                await _semaphore.WaitAsync(cancellationToken);
                 _items.AddRange(items);
-                if (_clock.GetCurrentDateTimeUtc() >= _nextFlush) { 
-                  Flush();
+                if (_clock.GetCurrentDateTimeUtc() >= _nextFlush)
+                {
+                    await FlushAsync(cancellationToken, waitForSemaphore: false);
                 }
             }
+            finally
+            {
+                _semaphore.Release();
+            }
         }
 
         /// <inheritdoc />
-        public void Flush()
+        public void Flush() => Flush(waitForSemaphore: true);
+
+        /// <inheritdoc />
+        public async Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken)) => 
+            await FlushAsync(cancellationToken, waitForSemaphore: true);
+
+        /// <summary>
+        /// Flush the buffer.
+        /// </summary>
+        /// <param name="waitForSemaphore">If true the flush will wait for the semaphore.</param>
+        private void Flush(bool waitForSemaphore)
+        {
+            IList<T> old;
+            if (waitForSemaphore) _semaphore.Wait();
+            old = _items;
+            _items = new List<T>();
+            _nextFlush = DateTime.UtcNow.Add(_waitTime);
```

Comment:
I'd personally write this as `DateTime.UtcNow + _waitTime` to make it clearer that `Add` doesn't mutate, but no biggie.